c7n/registry.py

```diff
@@ -45,28 +45,47 @@ class PluginRegistry(object):
 
     """
 
+    event_times = ('register', 'unregister', 'final')
+
     def __init__(self, plugin_type):
         self.plugin_type = plugin_type
         self._factories = {}
+        self._events = {x: [] for x in self.event_times}
+
+    def add_event(self, time, func):
+        if time not in self.event_times:
+            raise ValueError('Available hook times are: {}'.format(', '.join(self.event_times)))
+
+        self._events[time].append(func)
 
     def register(self, name, klass=None):
         # invoked as function
         if klass:
+            self.run_events('register', name)
             klass.type = name
             self._factories[name] = klass
             return klass
 
         # invoked as class decorator
         def _register_class(klass):
+            self.run_events('register', name)
             self._factories[name] = klass
             klass.type = name
             return klass
         return _register_class
 
     def unregister(self, name):
+        self.run_events('unregister', name)
         if name in self._factories:
             del self._factories[name]
 
+    def run_final_events(self):
+        self.run_events('final', None)
+
+    def run_events(self, category, key):
+        for event in self._events[category]:
+            event(self, key, self._factories.keys())
```

Comment:
s/event/subscriber

also we should be distinguishing on collection vs registration events on params. registratin events should get (event_name, registry_key, and value) .. collection/registry events should get (event_name, registry).